### PR TITLE
Fixes #20079 - SSL secured and verified Pg connection

### PR DESCRIPTION
**Pull request by mbacovsky**
To setup DB with SSL and verification use params:

  db_sslmode => 'verify-full',
  db_root_cert => 'ca_bundle.pem'

Default DB sslmode is 'prefer' - non-verified SSL with fallback to
non-SSL. To use SSL secured connection with CA verification sslmode
needs to be set to either 'verify-ca' or 'verify-full'. Underlying libpg
uses DB root cert stored at '~/.postgresql/root.crt' which is in case of
Foreman at '/usr/share/foreman/.postgresql/root.crt'. There is no way to
setup different path (besides using env vars). System CA trust is not
supported. The cert needs to be real file as links are not allowed too.

For more details see SSL support in libpg:
  https://www.postgresql.org/docs/9.2/static/libpq-ssl.html

### CODE REVIEW

**Comment by mbacovsky** [comment]
Redmine ticket updated accordingly.

Question for reviewer: Is the `prefer` secure default sslmode? It implicitly uses non-verified SSL connection without warning to the user. Wouldn't be 'disabled` better default forcing the user to choose the sslmode consciously?

File: manifests/database/postgresql.pp [hunk 1]
```diff
@@ -10,6 +10,22 @@
     default => postgresql_password($::foreman::db_username, $::foreman::db_password),
   }
 
+  if $::foreman::db_root_cert {
+    $pg_cert_dir = "${::foreman::app_root}/.postgresql/"
+
+    file { $pg_cert_dir:
+      ensure => 'directory',
```
Comment (ekohl) [inline manifests/database/postgresql.pp]: You should manage owner and permission here as well. I've learned that it's better to be explicit about that than to rely on defaults. There could be a weird umask for example.

File: manifests/init.pp [hunk 2]
```diff
@@ -225,6 +227,7 @@
   Optional[String] $db_username = $::foreman::params::db_username,
   Optional[String] $db_password = $::foreman::params::db_password,
   Optional[String] $db_sslmode = 'UNSET',
+  Optional[String] $db_root_cert = 'UNSET',
```
Comment (ekohl) [inline manifests/init.pp]: I dislike the ```UNSET```. It used to be needed in puppet versions but now I think we can just use ```undef```.

File: manifests/database/postgresql.pp [hunk 3]
```diff
@@ -10,6 +10,25 @@
     default => postgresql_password($::foreman::db_username, $::foreman::db_password),
   }
 
+  if $::foreman::db_root_cert {
+    $pg_cert_dir = "${::foreman::app_root}/.postgresql/"
```
Comment (ekohl) [inline manifests/database/postgresql.pp]: Nit: this trailing slash isn't needed.

**Comment by mbacovsky** [comment]
@ekohl, updated. I gave the sslmode conditioned by presence of the root cert a try. Does it make sense and is it worth it?

**Comment by mmoll** [comment]
This is looking good, @mbacovsky I presume you tested this already with verification using the Katello certificate factory?

**Comment by mbacovsky** [comment]
@mmoll I've created standalone CA cert chain to sign the pg server certificate. My guess was this will better simulate the real deployment where the Foreman and remote Pg server will have certs signed by common corporate CA or the Pg server CA will be totally independent. 

Anyway the patch just installs the pg server CA cert so it shouldn't depend on where the cert comes from.

**Comment by mmoll** [comment]
fair enough, merged, díky @mbacovsky!

**Pull request merged**